### PR TITLE
fix: [PL-42171]: Sending heartbeat should not be dependent on acquireTask value

### PR DESCRIPTION
**Pull request by VikasMaddukuriHarness**
## Describe your changes
When shutDown hook is received in delegate, we stop acquiring new tasks by marking the atomic value acquireTask to false. We were using this value to whether or not send out heartbeat. So incase where the task is being executed and the shutdown hook is triggered, the task will still be running in delegate but the heartbeat is not sent and this causes the delegate to be marked as deleted in manager side leading to marking of all the delegateTaks which are under the current delegate as UN_ASSIGNED. But this is not the case because the task is actually being executed and the websocket connection was still open.
Only because the heartbeat was not received for more than 5 minutes, manager marks the delegate as disconnected and mark all the tasks as UN_ASSIGNED.

This pr fixes this by removing the dependency of acquireTask field on sending heartbeat.

Testing:
1. Created a delegate task VALIDATE_SECRET_MANAGER_CONFIGURATION which waits for runs for more than 60 seconds. 
After the changes, we need to see the heartbeat still being sent even though the shutdown hook is called and task should be successful. Attached SS for the same - 
Screenshot 2023-11-08 at 3 57 11 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55165)

### CODE REVIEW

**Review by xingchi-jin — APPROVED** [review]

File: 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java [hunk 1]
```diff
@@ -1905,7 +1905,7 @@ private void sendHeartbeat(final DelegateParamsBuilder builder) {
   }
 
   private void sendHttpHeartbeat(DelegateParamsBuilder builder) {
-    if (!shouldContactManager() || !acquireTasks.get() || frozen.get()) {
```
Comment (xingchi-jin) [inline 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java:1908]: the acquireTasks set to false in two places.
one is the shutdown hook - covered,
the other place is self-destruct.

The code is used in both immutable and legacy. let's make sure the second case is also covered and no regression

Comment (VikasMaddukuriHarness) [reply 260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java:1908]: shouldContactManager method actually checks for the selfDestruct is true or not. 
So if self-destruct signal is received, it stops sending heartbeat.

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Comment by VikasMaddukuriHarness** [comment]
trigger messagecheck

**Pull request merged**